### PR TITLE
Turnkey-LLM (aka lemonade)

### PR DESCRIPTION
**Pull request by jeremyfowers**
This update adds the `turnkey-llm` CLI and `turnkeyml.llm` APIs, which provide LLM-specific tools for prompting, serving, and accuracy tests.

This initial release has device/runtime support for:
- Huggingface LLMs with CPU and CUDA backends
- OnnxRuntime-GenAI with DirectML backend
- RyzenAI-Transformers backend

The tools provided are:
- `huggingface-load`: load LLM from huggingface checkpoint
- `oga-load`: load LLM using OnnxRuntime-GenAI onto DirectML device
- `ryzenai-npu-load`: load LLM onto RyzenAI NPU device
- `llm-prompt`: chat with your LLM
- `serve`: launch a websocket server for your LLM
- `accuracy-mmlu` and `accuracy-perplexity`: measure the performance of your LLM on the target backend

Learn more about the features and usage in the [README.md](https://github.com/onnx/turnkeyml/blob/abe0a39619cccb69746c4757e893a7104fe5b9f8/src/turnkeyml/llm/README.md) at src/turnkeyml/llm/README.md.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -44,10 +44,46 @@
...
+            "numpy==1.26.4",
+            "uvicorn[standard]",
+        ],
+    },
     classifiers=[],
     entry_points={
         "console_scripts": [
             "turnkey=turnkeyml:turnkeycli",
+            "turnkey-llm=turnkeyml.llm:lemonadecli",
+            "lemonade=turnkeyml.llm:lemonadecli",
```
Comment (danielholanda) [inline setup.py:86]: Is the plan to keep multiple entry points for the same functionality long-term?

File: src/turnkeyml/version.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = "4.0.0"
+__version__ = "4.0.1"
```
Comment (danielholanda) [inline src/turnkeyml/version.py:1]: MINOR: Incremented when adding new features or functionality in a backward-compatible manner.
```suggestion
__version__ = "4.1.0"
```

File: src/turnkeyml/llm/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
+# Turnkey-LLM
```
Comment (danielholanda) [inline src/turnkeyml/llm/README.md:1]: This PR is a major contribution! Some of this info definitely deservers to be part of the main README.md file of the repo.
(Could be a separate PR)

File: src/turnkeyml/llm/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+1. Clone: `git clone https://github.com/onnx/turnkeyml.git`
+1. Create and activate a conda environment:
+    1. `conda create -n tk-llm python=3.10`
+    1. `conda activate tk-llm`
+1. `cd turnkeyml` (where `turnkeyml` is the repo root of your TurnkeyML clone)
+1. Install lemonade: `pip install -e .[llm]`
+    - or `pip install -e .[llm-oga-dml]` if you want to use `onnxruntime-genai` (see [OGA](#install-onnxruntime-genai))
+1. `lemonade -h` to explore the LLM tools
+
```
Comment (danielholanda) [inline src/turnkeyml/llm/README.md:25]: Those instructions are in src/turnkeyml/llm, but they must be run from the root folder of the repo, which is confusing. Please clarify that.

File: src/turnkeyml/llm/README.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+You can also set the `--device` argument in `huggingface-load` to load your LLM on a different device.
+
+Run `lemonade huggingface-load -h` and `lemonade llm-prompt -h` to learn more about those tools.
+
+## Accuracy
+
+To measure the accuracy of an LLM using MMLU, try this:
+
+`lemonade -i facebook/opt-125m huggingface-load mmlu-accuracy --tests management`
```
Comment (danielholanda) [inline src/turnkeyml/llm/README.md:48]: Installed using `pip install -e .[llm-oga-dml]` and can successfully run `lemonade -i facebook/opt-125m huggingface-load llm-prompt -p "Hello, my thoughts are"` but not `lemonade -i facebook/opt-125m huggingface-load mmlu-accuracy --tests management`

```
Error: unrecognized arguments: mmlu-accuracy --tests management. If `mmlu-accuracy --tests management` was intended to invoke a tool, please run `turnkey -h` and check the spelling and availability of that tool.

▄██████████████▄▐█▄▄▄▄█▌
██████▌▄▌▄▐▐▌███▌▀▀██▀▀
████▄█▌▄▌▄▐▐▌▀███▄▄█▌
▄▄▄▄▄██████████████
```

Is mmlu only available when something else is installed? If so, please update the instructions.

**Review by danielholanda — APPROVED** [review]

File: src/turnkeyml/llm/README.md (hunk 5 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md:48]: its accuracy-mmlu 😓

File: src/turnkeyml/version.py (hunk 2 above)
Comment (jeremyfowers) [reply src/turnkeyml/version.py:1]: That's not how we define MINOR in https://github.com/onnx/turnkeyml/blob/main/docs/versioning.md

File: src/turnkeyml/llm/README.md (hunk 3 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md:1]: Good idea!

File: src/turnkeyml/llm/README.md (hunk 4 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md:25]: Good catch

File: setup.py (hunk 1 above)
Comment (jeremyfowers) [reply setup.py:86]: Yeah. Can't let go of the classics.

File: src/turnkeyml/llm/README.md (hunk 5 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md:48]:
```
(tk-llm) PS C:\work\turnkeyml> lemonade -i facebook/opt-125m huggingface-load accuracy-mmlu --tests management

Building "facebook_opt-125m"
    ✓ Loading Huggingface checkpoint   
    ✓ Measuring accuracy with MMLU   

facebook/opt-125m:
        <built-in function input> (executed 1x)
                Parameters:     125,239,296 (477.75 MB)
                Build dir:      C:\Users\jefowers/.cache/lemonade\facebook_opt-125m
                Status:         Successful build!
                                Dtype:  float32 
                                Mmlu Management Accuracy:       19.417 %
```

Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md]: Fixed in https://github.com/onnx/turnkeyml/pull/225/commits/4b2c85e948e603a0f80d0b8deebd584befdc6695

File: src/turnkeyml/llm/README.md (hunk 4 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md]: Updated https://github.com/onnx/turnkeyml/pull/225/commits/4b2c85e948e603a0f80d0b8deebd584befdc6695

File: src/turnkeyml/llm/README.md (hunk 3 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/README.md:1]: Added in https://github.com/onnx/turnkeyml/pull/225/commits/e7d7808b4b4135f6d76fbeb1f17ca6571b7cbaa2

File: src/turnkeyml/llm/tools/ryzenai_npu/ryzenai_npu.py [hunk 6]
```diff
@@ -137,6 +144,21 @@ def run(self, state: State, input: str = "", device=None) -> State:
...
+        ):
+            model_name = checkpoint.replace("meta-llama/", "")
+            algorithm = "awqplus"
+            flash_attention_plus = False
+            trust_remote_code = False
+            CausalLMModel = LlamaModelEval
+            LMTokenizer = PreTrainedTokenizerFast
+            quantized_model_path = os.path.join(
+                quantized_models_path,
+                f"quantized_llama-2-7b-chat_w{w_bit}_g{group_size}_{algorithm}.pth",
```
Comment (kovtcharov) [inline src/turnkeyml/llm/tools/ryzenai_npu/ryzenai_npu.py:159]: should string be `quantized_llama-3-8b`?

**Review by kovtcharov — APPROVED** [review]

File: src/turnkeyml/llm/tools/ryzenai_npu/ryzenai_npu.py (hunk 6 above)
Comment (jeremyfowers) [reply src/turnkeyml/llm/tools/ryzenai_npu/ryzenai_npu.py]: yup, fixed https://github.com/onnx/turnkeyml/pull/225/commits/c05d6295d931f9811858842b7109dd798369f0c2
